gnovm/stdlibs/strings/replace.gno

```diff
@@ -0,0 +1,620 @@
+		}
+	}
+	if last != len(s) {
+		nw, err := sw.WriteString(s[last:])
+		n += nw
+		if err != nil {
+			return n, err
+		}
+	}
+	return n, nil
+}
+
+// byteStringReplacer is the implementation that's used when all the
+// "old" values are single ASCII bytes but the "new" values vary in size.
+type byteStringReplacer struct {
+	// replacements contains replacement byte slices indexed by old byte.
+	// A nil []byte means that the old byte should not be replaced.
+	replacements [256][]byte
+	// toReplace keeps a list of bytes to replace. Depending on length of toReplace
+	// and length of target string it may be faster to use Count, or a plain loop.
+	// We store single byte as a string, because Count takes a string.
+	toReplace []string
+}
+
+// countCutOff controls the ratio of a string length to a number of replacements
+// at which (*byteStringReplacer).Replace switches algorithms.
+// For strings with higher ration of length to replacements than that value,
+// we call Count, for each replacement from toReplace.
+// For strings, with a lower ratio we use simple loop, because of Count overhead.
+// countCutOff is an empirically determined overhead multiplier.
+// TODO(tocarip) revisit once we have register-based abi/mid-stack inlining.
+const countCutOff = 8
+
+func (r *byteStringReplacer) Replace(s string) string {
+	newSize := len(s)
+	anyChanges := false
+	// Is it faster to use Count?
+	if len(r.toReplace)*countCutOff <= len(s) {
+		for _, x := range r.toReplace {
+			if c := Count(s, x); c != 0 {
+				// The -1 is because we are replacing 1 byte with len(replacements[b]) bytes.
+				newSize += c * (len(r.replacements[x[0]]) - 1)
+				anyChanges = true
+			}
+
+		}
+	} else {
+		for i := 0; i < len(s); i++ {
+			b := s[i]
+			if r.replacements[b] != nil {
+				// See above for explanation of -1
+				newSize += len(r.replacements[b]) - 1
+				anyChanges = true
+			}
+		}
+	}
+	if !anyChanges {
+		return s
+	}
+	buf := make([]byte, newSize)
+	j := 0
+	for i := 0; i < len(s); i++ {
+		b := s[i]
+		if r.replacements[b] != nil {
+			j += copy(buf[j:], r.replacements[b])
+		} else {
+			buf[j] = b
+			j++
+		}
+	}
+	return string(buf)
+}
+
+func (r *byteStringReplacer) WriteString(w io.Writer, s string) (n int, err error) {
+	sw := getStringWriter(w)
+	last := 0
+	for i := 0; i < len(s); i++ {
+		b := s[i]
+		if r.replacements[b] == nil {
+			continue
+		}
+		if last != i {
+			nw, err := sw.WriteString(s[last:i])
+			n += nw
+			if err != nil {
+				return n, err
+			}
+		}
+		last = i + 1
+		nw, err := w.Write(r.replacements[b])
+		n += nw
+		if err != nil {
+			return n, err
+		}
+	}
+	if last != len(s) {
+		var nw int
+		nw, err = sw.WriteString(s[last:])
+		n += nw
+	}
+	return
+}
+
+// Custom code: these test functions are normally placed on export_test.go file but it seems like we do not recognize
+// methods declared on test files as part of the structure. For example:
+// Replacer.printTrie() was not recognized so I added here directly
```

Comment:
wdym? what was the problem?

Reply:
If I deplace
```go
func (r *Replacer) PrintTrie() string {
	r.buildOnce()
	gen := r.r.(*genericReplacer)
	return gen.printNode(&gen.root, 0)
}

func (r *genericReplacer) printNode(t *trieNode, depth int) (s string) {
	if t.priority > 0 {
		s += "+"
	} else {
		s += "-"
	}
	s += "\n"

	if t.prefix != "" {
		s += Repeat(".", depth) + t.prefix
		s += r.printNode(t.next, depth+len(t.prefix))
	} else if t.table != nil {
		for b, m := range r.mapping {
			if int(m) != r.tableSize && t.table[m] != nil {
				s += Repeat(".", depth) + string([]byte{byte(b)})
				s += r.printNode(t.table[m], depth+1)
			}
		}
	}
	return
}
```
To the **export_test.go** file where it is originally on Go native lib. The method is not recognized by the test and I got the following error:
`.: test pkg: panic: strings_test/replace_test.gno:406:10: missing field PrintTrie in *strings.Replacer:`

Reply:
maybe there is a bug I could investigate, at least a mismatch with Go behavior

Reply:
I think I get it.

When we do the import on "strings" in the testing context here, from the package `strings_test`, we don't load what is declared in the test files. (This is different from Go's behaviour, please open up an issue)

Can you put the test that uses `PrintTrie` (`TestGenericTrieBuilding` in a separate test file, together with this `PrintTrie` method, declared in `package strings`?

This way we can use PrintTrie without actually making it a publicly available method (it shouldn't be.)

Reply:
@thehowl fixed on b31ccc2f38674b254dd0927ae0cc70b42faed90f
Also New issue opened #2888